### PR TITLE
[prolly] Float keyRange increment bug

### PR DESCRIPTION
**Pull request by max-hoffman**
Incrementing the `[n, n+1)` key range is a lot faster than a binary search with a tuple comparison callback. But it is subject to at least two edge cases where `(n+1)` is not a valid stop range: (1) `n+1 == n`, because of precision loss, and (2) `n+1 < n`, because of overflow.

I added a series of GMS tests here: https://github.com/dolthub/go-mysql-server/pull/2554. I couldn't find a DECIMAL failure case, I think DECIMAL always encodes a valid 1's place, and is not subject to overflow AFAICT.

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>7fa23ad</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>7fa23ad</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/enginetest/dolt_queries.go [hunk 1]
```diff
@@ -735,6 +735,19 @@ var DoltScripts = []queries.ScriptTest{
 			},
 		},
 	},
+	{
```
Comment (timsehn) [inline go/libraries/doltcore/sqle/enginetest/dolt_queries.go:738]: Maybe a bit of a battery of float tests as well? I'm concerned about our coverage since this was only caught in fuzzer testing.

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>de5000c</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>de5000c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>8b3f6fc</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>8b3f6fc</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/store/prolly/tuple_range.go [hunk 2]
```diff
@@ -167,18 +173,161 @@ func (r Range) Matches(t val.Tuple) bool {
...
+	if n < 0 {
+		// ex: range scan
+		return nil, false
+	}
+
+	for _, typ := range r.Desc.Types[n+1:] {
+		if !typ.Nullable {
+			// this is checked separately because fulltext descriptors
+			// do not match field lengths sometimes
+			// todo: why?
```
Comment (nicktobey) [inline go/store/prolly/tuple_range.go:222]: Can you create an issue for this and link to it?

File: go/store/prolly/tuple_range.go [hunk 3]
```diff
@@ -167,18 +173,161 @@ func (r Range) Matches(t val.Tuple) bool {
...
+		if !ok {
+			return nil, false
+		}
+		tb.PutUint64(n, v+1)
+	case val.Float32Enc:
+		v, ok := r.Desc.GetFloat32(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutFloat32(n, v+1)
```
Comment (nicktobey) [inline go/store/prolly/tuple_range.go:306]: Instead of adding 1 here (and elsewhere), we could find the next highest representable float. This would be useful both for very high values that don't have enough precision to add 1, but also very low values where adding 1 is an enormous range.

File: go/store/prolly/tuple_range.go [hunk 4]
```diff
@@ -167,18 +173,161 @@ func (r Range) Matches(t val.Tuple) bool {
...
+		if !ok {
+			return nil, false
+		}
+		tb.PutDecimal(n, v.Add(decimal.New(1, 0)))
+	default:
+		return nil, false
+	}
+	stop := tb.Build(pool)
+	if r.Desc.Compare(r.Tup, stop) >= 0 {
+		// |stop| has to be strictly greater than |start|
```
Comment (nicktobey) [inline go/store/prolly/tuple_range.go:324]: It's not clear from this comment that we're checking for overflow/lack of precision here. Maybe clarify?

**Review by nicktobey — APPROVED** [review]

File: go/store/prolly/tuple_range.go (hunk 2 above)
Comment (max-hoffman) [reply go/store/prolly/tuple_range.go:222]: yep https://github.com/dolthub/dolt/issues/8029

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>731b74b</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>731b74b</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>3124cf1</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>3124cf1</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>
